config/Config.java

```diff
@@ -0,0 +1,81 @@
+package grakn.simulation.config;
+
+import grakn.simulation.agents.base.AgentRunner;
+
+import java.util.List;
+import java.util.function.Function;
+
+public class Config {
+    private List<Agent> agents;
+    private TraceSampling traceSampling;
+
+    public List<Agent> getAgents() {
+        return agents;
+    }
+
+    public void setAgents(List<Agent> agents) {
+        this.agents = agents;
+    }
+
+    public TraceSampling getTraceSampling() {
+        return traceSampling;
+    }
+
+    public void setTraceSampling(TraceSampling traceSampling) {
+        this.traceSampling = traceSampling;
+    }
+
+    public static class TraceSampling {
+        private Schema.SamplingFunction function;
+        private Integer arg;
+
+        public Function<Integer, Boolean> getSamplingFunction() {
+            return Schema.SamplingFunction.applyArg(function, arg);
+        }
+
+        public void setFunction(String function) {
+            this.function = Schema.SamplingFunction.getByName(function);
+        }
+
+        public Integer getArg() {
+            return arg;
+        }
+
+        public void setArg(Integer arg) {
+            this.arg = arg;
+        }
+    }
+
+    public static class Agent {
+        private Schema.AgentMode agentMode;
+        private AgentRunner<?> runner;
+
+        public void setName(String name) {
+            this.runner = Schema.AGENTS.get(name);
```

Comment:
I think you should refactor this to only set a `String name` and then construct the full `AgentRunner` in `getRunner()`, which should be renamed `buildRunner()` to be clearer. This way we can isolate the runner construction logic to the `buildRunner()` method, and it also means we would have all runner config available before constructing the runner, so we could have `trace` as a constructor parameter.

Reply:
This would avoid the need for the construction and config logic of the runner to be in simulation main.

Reply:
I like it, will implement

Reply:
I've tried this and actually it looks worse to me - we still can't pass `trace` as a constructor parameter without passing around to both the runner and the agent classes and then instantiating the runner with `trace` and the agent class.

We also don't want to move the runner building logic into the config classes. The idea is that these classes are dumb holders of information, largely driven in their design by the requirements from snakeyaml. Even if we did, we still wouldn't be able to move other agent config checks inside this `build` method without obfuscating which agents are being built and in which way.